Fix some warnings

Holzhaus opened the pull request:
This fixes a bunch of warnings.

| Warning                             | gcc 9.2.0 (master) | gcc 9.2.0 (this branch) | clang 9.0.0 (master) | clang 9.0.0 (this branch)
| ----------------------------------- | ------------ | ----------------- | -------------- | -------------------
| `[-Wclass-memaccess]`               | 1            |                   |                |
| `[-Wdeprecated-copy]`               | 1            | 1                 |                |
| `[-Wdeprecated-declarations]`       | 54           | 18                | 42             | 18
| `[-Winconsistent-missing-override]` |              |                   | 1237           |
| `[-Wrestrict]`                      | 13           | 13                |                |
| `[-Wunused-const-variable]`         |              |                   | 2              |
| `[-Wunused-private-field]`          |              |                   | 6              |

Number of warnings were determined using `$ grep -Po "^src/.*\K\[-W.*\]" stderr.log | sort | uniq -c`.

**EDIT:** Updated the stats above after some new commits (and now using SCons instead of CMake). This also fixes the `-Wclass-memaccess` warning now.

**EDIT 2:** Pushed some more changes, updated the stats above and also checked using clang. To check what warnings are actually generated, I removed the `-Wno-*` flags locally:
```diff
$ git diff
diff --git a/build/depends.py b/build/depends.py
index 3cab15e57b..16f8ba9930 100644
--- a/build/depends.py
+++ b/build/depends.py
@@ -1368,21 +1368,7 @@ class MixxxCore(Feature):
             build.env.Append(CCFLAGS='-Wall')
             build.env.Append(CCFLAGS='-Wextra')

-            if build.compiler_is_gcc and build.gcc_major_version >= 9:
-                # Avoid many warnings from GCC 9 about implicitly defined copy assignment
-                # operators that are deprecated for classes with a user-provided copy
-                # constructor. This affects both Qt 5.12 and Mixxx.
-                build.env.Append(CXXFLAGS='-Wno-deprecated-copy')
-
             if build.compiler_is_clang:
-                # Quiet down Clang warnings about inconsistent use of override
-                # keyword until Qt fixes qt_metacall.
-                build.env.Append(CCFLAGS='-Wno-inconsistent-missing-override')
-
-                # Do not warn about use of the deprecated 'register' keyword
-                # since it produces noise from libraries we depend on using it.
-                build.env.Append(CCFLAGS='-Wno-deprecated-register')
-
                 # Warn about implicit fallthrough.
                 build.env.Append(CCFLAGS='-Wimplicit-fallthrough')
```

Holzhaus committed:
widget/wtracktableview: Fix deprecated QTableView::sortByColumn(int)

Holzhaus committed:
library/dlgcoverartfullsize: Remove deprecated QApplication::desktop()

Use getPrimaryScreen() helper function from util/compatibility.h.
If no screen is found, fall back to 800x600px.

Holzhaus committed:
library/basesqltablemodel: Replace qStableSort with std::stable_sort

Holzhaus committed:
mixxx: Remove deprecated QApplication::desktop()

Use getPrimaryScreen() helper function from util/compatibility.h.
If no screen was found, the window is not moved.

Holzhaus committed:
preferences/dialog/dlgpreferences: Remove deprecated QApplication::desktop()

Use getPrimaryScreen() helper function from util/compatibility.h.
If no screen is found, fall back to 800x600px.

Holzhaus committed:
track/beatmap: Replace qLower/UpperBound with std::lower/upper_bound

Holzhaus committed:
track/beatutil: Replace qLowerBound with std::lower_bound

uklotzde commented on src/library/dlgcoverartfullsize.cpp:
```diff
@@ -130,8 +132,9 @@ void DlgCoverArtFullSize::slotCoverFound(const QObject* pRequestor,
-        const QSize availableScreenSpace =
-            QApplication::desktop()->availableGeometry().size() * 0.9;
+        const QList<QScreen*> screens = QGuiApplication::screens();
+        DEBUG_ASSERT(!screens.isEmpty());
```
This list should never be empty, but we cannot rely on this assumption. The Qt docs do not promise anything. Better use VERIFY_OR_DEBUG_ASSERT with a warning (or critical?) log message and either fall back to some default behavior/settings or abort the operation.

uklotzde replied:
Maybe encapsulated in a getPrimaryScreen() utility function that reveals the intention instead of just using a low-level function like QList::first().

Holzhaus replied:
> Maybe encapsulated in a getPrimaryScreen() utility function that reveals the intention instead of just using a low-level function like QList::first().

I re-checked the Qt docs and found [QGuiApplication's primaryScreen property](https://doc.qt.io/qt-5/qguiapplication.html#primaryScreen-prop). This works for Qt >= 5.6. Is mixxx compatible with older Qt5 versions? Also, this property isn't static, so we'd need to pass around a reference to the QtGuiApplication instance.

uklotzde replied:
Currently the minimum supported version is Qt 5.4.

Holzhaus replied:
Since `QScreen` has no public constructor, I decided against `getPrimaryScreen` and used `VERIFY_OR_DEBUG_ASSERT`. Please have a look at the respective commits and tell if the implementation is okay.

Holzhaus committed:
util/fifo: Remove memset on object ptr to fix class-memaccess warning

Judging from a superficial look at the pa_ringbuffer implementation,
this memset call shouldn't be necessary anyway.

Holzhaus committed:
library/autodj/autodjfeature: Replace deprecated QModelIndex::child() call

Holzhaus committed:
library/dao/playlistdao: Use QT_VERSION_CHECK to fix deprecation warning

Holzhaus committed:
library/itunes/itunesfeature: Replace deprecated QSqlError::number call

Holzhaus committed:
util/desktophelper: Replace deprecated QDir assignment with QDir::setPath

Holzhaus committed:
widget/woverview: Use QT_VERSION_CHECK to fix deprecation warning

Holzhaus commented:
All remaining deprecated declaration warnings are caused by the use of `QSignalMapper`. I didn't really look into it, but [apparently `QSignalMapper` can be replaced by using lambda expressions.](https://github.com/qbittorrent/qBittorrent/commit/3dd969df643f68b465e10f29a4c8611607ca9968)

uklotzde commented on src/library/dao/playlistdao.cpp:
```diff
@@ -974,8 +974,22 @@ void PlaylistDAO::shuffleTracks(const int playlistId, const QList<int>& position
+        #pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+        #pragma clang diagnostic push
+        #pragma clang diagnostic ignored "-Wdeprecated-declarations"
         newPositions.swap(newPositions.indexOf(trackAPosition),
```
How about guarding the new code for Qt 5.13 behind a QT_VERSION_CHECK and keeping the old code in the #else branch? If the differences are limited to a very close scope this is the preferred way to go.

Holzhaus replied:
Done.

uklotzde commented on src/preferences/dialog/dlgpreferences.cpp:
```diff
@@ -406,9 +406,15 @@ void DlgPreferences::onShow() {
+        qWarning() << "No screen was found, assuming screen size 800x600px.";
+        screenSpace = QSize(800, 600);
+    } else {
+        screenSpace = screens.first()->geometry().size();
```
QList<Screen*>::first() is just one possible strategy for selecting the primary screen from this list. We can still hide it behind a one-line, inline utility function `Screen* getPrimaryScreen(QList<Screen*> screens)`, maybe in compatibility.h.

Holzhaus replied:
Done in https://github.com/mixxxdj/mixxx/pull/2350/commits/8c1159e2e11186f6b289f7395c147e3d3976907b.

uklotzde commented on src/widget/woverview.cpp:
```diff
@@ -1029,7 +1029,21 @@ void WOverview::paintText(const QString& text, QPainter* pPainter) {
+    #pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+    #pragma clang diagnostic push
+    #pragma clang diagnostic ignored "-Wdeprecated-declarations"
     int textWidth = fm.width(text);
```
Use QT_VERSION_CHECK here, too?

Holzhaus replied:
Ok.

Holzhaus committed:
util/compatibility: Add QScreen* getPrimaryScreen() helper function

uklotzde commented:
Thanks for taking care!!! LGTM, waiting for CI builds to complete.

Holzhaus commented:
I just checked with CXX=clang, there are some other warnings as well. I'll look into them but feel free to merge, I can just open a new PR ;-)

Btw, do you want me to squash the primary screen related commits?

uklotzde commented:
Squashing is not required, but feel free to do so. I will merge this later when you consider it done.

daschuer commented:
I can confirm that this still builds without any warning on Ubuntu Xenial. Thank you very much.

Holzhaus committed:
Add missing override to fix clang's -Winconsistent-missing-override warnings

Holzhaus committed:
effects/lv2/lv2manifest: Remove unused private field m_isValid

Holzhaus committed:
engine/sidechain/enginenetworkstream: Remove unused private field m_pInputWorker

Holzhaus committed:
engine/sidechain/networkinputstreamworker: Remove unused private fields

Holzhaus committed:
preferences/dialog/dlgprefbroadcast: Remove unused const variable

Holzhaus committed:
soundio/sounddevicenetwork: Remove unused const variable kBufferFrames

Holzhaus committed:
soundio/sounddevicenetwork: Remove unused private fields

Holzhaus commented:
I squashed the commits and fixed some more warnings. The tests still work fine for me, but please test again. If every works fine, feel free to merge ;-)

uklotzde committed:
Delete redundant virtual keywords

uklotzde committed:
Delete remaining redundant virtual keywords

uklotzde commented:
Thank you so much, Jan! I've pushed 2 commits that fix the redundant `virtual` keywords. Ready to merge if the CI builds succeed.

Holzhaus commented:
Merge?

uklotzde commented:
Of course :+1:
